[`Tokenizer`] Fix slow and fast serialization

ArthurZucker opened the pull request:
# What does this PR do?

- sets the defaults for AddedToken instances where needed to match what is pushed to the hub
- sets the default for AddedToken to not strip left and right to match the fast tokenizers
- fixes the `added_tokens.json` file: a recent push made it save all the added tokens encoder, but it should only save the indexes greater than the vocab size for forward compatibility. 
- fixes the list of `additionnal_special_tokens` that were added twice / overwritten
- fixes `add_tokens` : if the added tokens is a string we check if it's not already in the added vocab instead of always defaulting to strip left or right.
- fixes saving: the added_tokens_decoder should not add a `"__type ":"AddedToken"` field to the added tokens otherwise the previous versions of transformers will try to load them.

fixes #26732, fixes #26775, fixes #26773, fixes #26768, fixes #26859

ArthurZucker committed:
fix

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-main

ArthurZucker committed:
last attempt

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26570). All of your documentation changes will be reflected on that endpoint.

ArthurZucker committed:
current work

ArthurZucker committed:
fix forward compatibility

ArthurZucker committed:
save all special tokens

ArthurZucker committed:
Merge branch 'fix-main' of github.com:ArthurZucker/transformers into fix-main

ArthurZucker committed:
current state

ArthurZucker committed:
revert additional changes

ArthurZucker committed:
updates

ArthurZucker committed:
remove tokenizer.model

ArthurZucker committed:
add a test and the fix

ArthurZucker committed:
nit

ArthurZucker committed:
revert one more break

ArthurZucker committed:
fix typefield issue

ArthurZucker committed:
quality

ArthurZucker committed:
more tests

ArthurZucker committed:
fix fields for FC

ArthurZucker committed:
Merge branch 'fix-main' of github.com:ArthurZucker/transformers into fix-main

ArthurZucker committed:
Merge branch 'fix-main' of github.com:ArthurZucker/transformers into fix-main

ArthurZucker committed:
more nits?

ArthurZucker committed:
Merge branch 'fix-main' of github.com:ArthurZucker/transformers into fix-main

ArthurZucker committed:
new additional changes

ArthurZucker committed:
how

ArthurZucker committed:
some updates

ArthurZucker committed:
simplify all

ArthurZucker committed:
more nits

ArthurZucker committed:
revert some things to original

ArthurZucker committed:
nice

ArthurZucker committed:
nits

ArthurZucker committed:
a small hack

ArthurZucker committed:
more nits

ArthurZucker committed:
ahhaha

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-main

ArthurZucker committed:
fixup

ArthurZucker committed:
update

ArthurZucker committed:
make test run on ci

ArthurZucker committed:
use subtesting

ArthurZucker committed:
update

ArthurZucker commented on .circleci/create_circleci_config.py:
```diff
@@ -127,6 +127,7 @@ def to_dict(self):
             },
         ]
         steps.extend([{"run": l} for l in self.install_steps])
+        steps.extend({"run": "pip install pytest-subtests"})
```
```suggestion
        steps.extend([{"run": "pip install pytest-subtests"}])
```

ArthurZucker committed:
Update .circleci/create_circleci_config.py

ArthurZucker committed:
updates

ArthurZucker committed:
Merge branch 'fix-main' of github.com:ArthurZucker/transformers into fix-main

ArthurZucker committed:
fixup

ArthurZucker committed:
nits

ArthurZucker committed:
replace typo

ArthurZucker committed:
fix the test

ArthurZucker committed:
nits

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-main

ArthurZucker committed:
update

ArthurZucker committed:
None max dif pls

ArthurZucker committed:
a partial fix

ArthurZucker committed:
had to revert one thing

ArthurZucker committed:
test the fast

ArthurZucker committed:
updates

ArthurZucker committed:
fixup

ArthurZucker committed:
and more nits

ArthurZucker committed:
more fixes

ArthurZucker committed:
update

ArthurZucker committed:
Oupsy :eye:

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-main

ArthurZucker committed:
nits

ArthurZucker committed:
fix marian

ArthurZucker committed:
on our way to heaven

LysandreJik reviewed (approved):
à priori looks good to me

ArthurZucker committed:
Update src/transformers/models/t5/tokenization_t5.py

Co-authored-by: Lysandre Debut <hi@lysand.re>

ArthurZucker committed:
fixup

leot13 commented on src/transformers/tokenization_utils_fast.py:
```diff
@@ -633,7 +633,8 @@ def _save_pretrained(
             )
-            added_vocab = self.get_added_vocab()
+            # make sure to be foward compatible
+            added_vocab = {tok: index for tok, index in self.added_tokens_encoder.items() if index > self.vocab_size}
```
```suggestion
            added_vocab = {tok: index for tok, index in self.added_tokens_encoder.items() if index >= self.vocab_size}
```

leot13 commented on src/transformers/tokenization_utils_base.py:
```diff
@@ -2498,7 +2525,8 @@ def _save_pretrained(
         )
-        added_vocab = self.get_added_vocab()
+        # the new get_added_vocab() also returns special tokens and tokens that have an index < vocab_size
+        added_vocab = {tok: index for tok, index in self.added_tokens_encoder.items() if index > self.vocab_size}
```
```suggestion
        added_vocab = {tok: index for tok, index in self.added_tokens_encoder.items() if index >= self.vocab_size}
```

leot13 reviewed (commented):
found this when saving idefics

ArthurZucker committed:
Update src/transformers/tokenization_utils_fast.py

Co-authored-by: Leo Tronchon <leo.tronchon@gmail.com>

ArthurZucker committed:
Update src/transformers/tokenization_utils_base.py

Co-authored-by: Leo Tronchon <leo.tronchon@gmail.com>

ArthurZucker committed:
fix phobert

ArthurZucker committed:
skip some things, test more

ArthurZucker committed:
nits

ArthurZucker committed:
fixup

ArthurZucker committed:
fix deberta

ArthurZucker committed:
update

ArthurZucker committed:
update

ArthurZucker committed:
more updates

ArthurZucker committed:
skip one test

ArthurZucker committed:
more updates

ArthurZucker committed:
fix camembert

ArthurZucker committed:
can't test this one

ArthurZucker committed:
more good fixes

ArthurZucker committed:
kind of a major update

- seperate what is only done in fast in fast init and refactor
- add_token(AddedToken(..., speicla = True)) ignores it in fast
- better loading

ArthurZucker committed:
fixup

ArthurZucker committed:
more fixups

ArthurZucker committed:
fix pegasus and mpnet

ArthurZucker committed:
remove skipped tests

ArthurZucker committed:
fix phoneme tokenizer if self.verbose

ArthurZucker committed:
fix individual models

ArthurZucker committed:
update common tests

ArthurZucker committed:
update testing files

ArthurZucker committed:
all over again

ArthurZucker committed:
nits

ArthurZucker committed:
skip test for markup lm

ArthurZucker committed:
fixups

ArthurZucker committed:
fix order of addition in fast by sorting the added tokens decoder

ArthurZucker committed:
proper defaults for deberta

ArthurZucker committed:
correct default for fnet

ArthurZucker committed:
nits on add tokens, string initialized to special if special

ArthurZucker committed:
skip irrelevant herbert tests

ArthurZucker committed:
main fixes

ArthurZucker committed:
update test added_tokens_serialization

ArthurZucker committed:
the fix for bart like models and class instanciating

ArthurZucker committed:
update bart

ArthurZucker committed:
nit!

ArthurZucker committed:
update idefix test

ArthurZucker committed:
fix whisper!

ArthurZucker committed:
some fixup

ArthurZucker committed:
fixups

ArthurZucker committed:
revert some of the wrong chanegs

ArthurZucker committed:
fixup

ArthurZucker committed:
fixup

LysandreJik reviewed (approved):
Feel free to merge when ready as seen offline with you

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-main

ArthurZucker committed:
skip marian

ArthurZucker committed:
skip the correct tests

ArthurZucker committed:
skip for tf and flax as well

tongyx361 commented:
I ran into the error below

```
Traceback (most recent call last):
  File ".../src/train_flash_attn_2.py", line 11, in <module>
    train()
  File ".../src/train.py", line 157, in train
    tokenizer = transformers.AutoTokenizer.from_pretrained(
  File ".../lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py", line 751, in from_pretrained
    return tokenizer_class.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
  File ".../lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2017, in from_pretrained
    return cls._from_pretrained(
  File ".../lib/python3.10/site-packages/transformers/tokenization_utils_base.py", line 2243, in _from_pretrained
    init_kwargs[key] = added_tokens_map.get(init_kwargs[key], init_kwargs[key])
TypeError: unhashable type: 'dict'
```

So I added some prints and get this intermediate values:

```
cls.SPECIAL_TOKENS_ATTRIBUTES: (list)['bos_token', 'eos_token', 'unk_token', 'sep_token', 'pad_token', 'cls_token', 'mask_token', 'additional_special_tokens']
added_tokens_map: (dict){'<unk>': AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True), '<s>': AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True), '</s>': AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True)}
init_kwargs: (dict){'add_bos_token': True, 'add_eos_token': False, 'bos_token': {'__type': 'AddedToken', 'content': '<s>', 'lstrip': False, 'normalized': True, 'rstrip': False, 'single_word': False}, 'clean_up_tokenization_spaces': False, 'eos_token': {'__type': 'AddedToken', 'content': '</s>', 'lstrip': False, 'normalized': True, 'rstrip': False, 'single_word': False}, 'legacy': None, 'model_max_length': 1024, 'pad_token': None, 'sp_model_kwargs': {}, 'unk_token': {'__type': 'AddedToken', 'content': '<unk>', 'lstrip': False, 'normalized': True, 'rstrip': False, 'single_word': False}, 'vocab_file': '.../tokenizer.model', 'tokenizer_file': '.../tokenizer.json', 'name_or_path': '...'}
key: (streos_token
init_kwargs[key]: (dict){'__type': 'AddedToken', 'content': '</s>', 'lstrip': False, 'normalized': True, 'rstrip': False, 'single_word': False}
```

According to the output, I made a fix, which seemed to work out:

```diff
# Passing AddedTokens and not strings to the class to prevent it from casting the string to a different AddedToken
for key in cls.SPECIAL_TOKENS_ATTRIBUTES & init_kwargs.keys():
    if added_tokens_map != {} and init_kwargs[key] is not None:
        if key != "additional_special_tokens":
            # >>> debug
            def print_info(name, obj):
                print(f"{name}: ({type(obj).__name__}){obj}")
            print_info("cls.SPECIAL_TOKENS_ATTRIBUTES", cls.SPECIAL_TOKENS_ATTRIBUTES)
            print_info("added_tokens_map", added_tokens_map)
            print_info("init_kwargs", init_kwargs)
            print_info("key", key)
            print_info("init_kwargs[key]", init_kwargs[key])
            # <<< debug
-            init_kwargs[key] = added_tokens_map.get(init_kwargs[key], init_kwargs[key])
+            init_kwargs[key] = added_tokens_map.get(key, init_kwargs[key]) # fix
```

ArthurZucker commented:
Could you share a reproducer? Would help me a lot as well!

tongyx361 commented:
> Could you share a reproducer? Would help me a lot as well!

Sorry that I'm too busy to do so right now 😭

But this only happened when I loaded the tokenizer of [Llemma-7B](https://huggingface.co/EleutherAI/llemma_7b) and tried to add a `"<pad>"` special token to it.

I hope this description could help you reproduce the error.

adamtwatson commented on src/transformers/tokenization_utils_base.py:
```diff
@@ -2231,36 +2191,59 @@ def _from_pretrained(
+        for key in cls.SPECIAL_TOKENS_ATTRIBUTES & init_kwargs.keys():
+            if added_tokens_map != {} and init_kwargs[key] is not None:
+                if key != "additional_special_tokens":
+                    init_kwargs[key] = added_tokens_map.get(init_kwargs[key], init_kwargs[key])
```
Hi @ArthurZucker,

I think this is supposed to be 
`init_kwargs[key] = added_tokens_map.get(key, init_kwargs[key])`
As `init_kwargs[key]` will return a dict in some cases, and cause unhashable type error:
```
site-packages/transformers/tokenization_utils_base.py", line 2242, in _from_pretrained
    init_kwargs[key] = added_tokens_map.get(init_kwargs[key], init_kwargs[key])
TypeError: unhashable type: 'dict'
```
Thanks.
